### PR TITLE
retain partial sync on fetch error

### PR DESCRIPTION
**Pull request by amfern**
If an error occurs during the gclient sync, we aim to retain the partial sync directory.
This helps prevent the need to re-fetch a large amount of source code only to encounter the same network issue again.
The sync directory can grow up to 14GB.
However, it will still be deleted when running bitbake -c cleanall.

To test the change, you can start a clean build of the `flutter-engine`. And before the do_fetch has finished, disconnect the network.(via control panel or any other means).
The do_fetch would fail, however the partial download would be retained in  `${DL_DIR}/gn`, so restarting the build with the network should continue fetching dependencies.

### CODE REVIEW

File: lib/gn.py [hunk 1]
```diff
@@ -112,6 +113,7 @@ def download(self, ud, d):
 
         uri = ud.url.split(";")[0]
 
+        ud.trying_to_fetch_with_gclient = True
```
Comment (jwinarske) [inline lib/gn.py:116]: Shouldn't this be set to False after a successful fetch?

**Review by jwinarske — CHANGES_REQUESTED** [review]

File: lib/gn.py (hunk 1 above)
Comment (amfern) [reply lib/gn.py]: Yeah, it's a good idea to put it back to False.

**Review by jwinarske — APPROVED** [review]

**Comment by jwinarske** [comment]
@amfern Looks good, just need some documentation on how to test this so it's maintained moving forward.

**Comment by amfern** [comment]
@jwinarske where would like me to add the documentation?

**Comment by jwinarske** [comment]
Document test strategy in the PR, then add a now somewhere in README.md

**Comment by WAKazuyoshiAkiyama** [comment]
LGTM

**Comment by jwinarske** [comment]
@amfern can you comment how you tested this?  That's all I need.

**Comment by amfern** [comment]
To test the change, you can start a clean build of the flutter-engine. And before the do_fetch has finished, disconnect the network.(via control panel or any other means).
The do_fetch would fail, however the partial download would be retained in ${DL_DIR}/gn, so restarting the build with the network should continue fetching dependencies.

**Pull request merged**